### PR TITLE
Secp256k1

### PR DESCRIPTION
**Pull request by larskuhtz**

### CODE REVIEW

File: src-secp256k1/Crypto/Secp256k1/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+
+-- | Not sure if this is faster. TODO benchmark it.
+--
+sqrtFp2 :: Fp -> (Bool, Fp)
+sqrtFp2 a = (r .^ 2 == a, r)
+  where
+    r = (a .^ ((pC + 1) `quot` 8)) .^ 2
+
+-- | Source of the follwoing algorith is th scp256k1 C library. We should
+-- benchmark it to check if it is actually faster.
```
Comment (larskuhtz) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:408]:
```suggestion
-- | Source of the follwoing algorith is th scp256k1 C library.
```

File: ethereum.cabal [hunk 2]
```diff
@@ -8,15 +8,16 @@ license: BSD-3-Clause
...
 author: Lars Kuhtz
 maintainer: lakuhtz@gmail.com
-copyright: Copyright (c) 2020-2022 Kadena LLC.
+copyright: Copyright (c) 2020-2023 Kadena LLC.
 category: Data
 tested-with:
-    GHC==9.2.4
-    GHC==9.0.2
-    GHC==8.10.7
+    , GHC==9.4.2
```
Comment (chessai) [inline ethereum.cabal:14]: add GHC 9.6.2?

Comment (larskuhtz) [reply ethereum.cabal:14]: That is included in #17, which can be merged as soon as chainweb-node supports aeson-2.2, which is currently blocked on https://github.com/haskell-servant/servant/pull/1695.

**Review by ak3n — APPROVED** [review]
Don't know much about the cryptography but overall the code LGTM.

File: src-secp256k1/Crypto/Secp256k1/Internal.hs [hunk 3]
```diff
@@ -0,0 +1,716 @@
...
+natToBytes l = B.pack . natToBytesInternal l
+
+natToBytesInternal
+    :: Int
+    -> Natural
+    -> [Word8]
+natToBytesInternal l n = go l n []
+  where
+    go 0 _ = id
+    go i m = let (a, b) = quotRem m 256 in go (i - 1) a . (int b :)
```
Comment (chessai) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:245]: not: put a bang on `i` to make GHC more likely to make this tail recursive/for the int to unpack

File: src-secp256k1/Crypto/Secp256k1/Internal.hs [hunk 4]
```diff
@@ -0,0 +1,716 @@
...
+--
+sqrtFp :: Fp -> (Bool, Fp)
+sqrtFp = sqrtFp1
+
+sqrtFp1 :: Fp -> (Bool, Fp)
+sqrtFp1 a = (r .^ 2 == a, r)
+  where
+    r = a .^ ((pC + 1) `quot` 4)
+
+-- | Not sure if this is faster. TODO benchmark it.
```
Comment (chessai) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:400]: There is a TODO for benchmarks here, but the above comment said that it's been benchmarked. Could you amend the comments to reconcile this?

File: src-secp256k1/Crypto/Secp256k1/Internal.hs [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+  where
+    u1 = e .* is
+    u2 = r .* is
+    rP = u1 .*. gC .+. u2 .*. pk -- TODO use Shamir's algorithm to make this more efficient
+    is = invM s
+
+-- | Recover public key for curve Secp256k1.
+--
+-- The chance that `secondKey` is needed (i.e. there are two solution for the x
+-- coordinate) is \(2^{128}\). Do we need to cover that case or can we just
```
Comment (chessai) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:650]: I think you mean the inverse of this.

File: test-secp256k1/Test/Crypto/Secp256k1/Internal.hs [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+        $ verify msgDigest r s pk === Right True
+  where
+    msgDigest = hashMsg @Sha3_256 $ B8.pack $ show msg
+
+prop_ecdsa_recover :: Int -> Property
+prop_ecdsa_recover msg = ioProperty $ do
+    (sk, pk) <- genKey
+    (r, s, isOddY, isSecondKey) <- sign sk msgDigest
+    return
+        $ classify isOddY "isOddY"
```
Comment (chessai) [inline test-secp256k1/Test/Crypto/Secp256k1/Internal.hs:169]: nice QuickCheck-fu 👍

**Review by chessai — APPROVED** [review]

**Comment by chessai** [comment]
Approved, but left a few comments of some small things I would like to see changed.

I'm not an expert on the cryptography here either, but the math makes sense to me. Reading the references helped.

File: src-secp256k1/Crypto/Secp256k1/Internal.hs (hunk 3 above)
Comment (larskuhtz) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:245]:
```suggestion
    go !i m = let (a, b) = quotRem m 256 in go (i - 1) a . (int b :)
```

File: src-secp256k1/Crypto/Secp256k1/Internal.hs [hunk 7]
```diff
@@ -0,0 +1,716 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
Comment (larskuhtz) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:1]:
```suggestion
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE BangPatterns #-}
```

File: src-secp256k1/Crypto/Secp256k1/Internal.hs (hunk 4 above)
Comment (larskuhtz) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:400]:
```suggestion
-- | Not sure if this is faster.
```

File: src-secp256k1/Crypto/Secp256k1/Internal.hs (hunk 5 above)
Comment (larskuhtz) [inline src-secp256k1/Crypto/Secp256k1/Internal.hs:650]:
```suggestion
-- coordinate) is one in \(2^{128}\). Do we need to cover that case or can we just
```

**Review by chessai — APPROVED** [review]

**Pull request merged**